util/src/main/java/com/google/cloud/hadoop/util/RetryHttpInitializer.java

```diff
@@ -60,119 +58,124 @@
-      errorCodeHandler.setBackOffRequired(
-          response ->
-              BASE_HTTP_BACKOFF_REQUIRED.isRequired(response)
-                  || response.getStatusCode() == HTTP_SC_TOO_MANY_REQUESTS);
-      if (sleeperOverride != null) {
-        errorCodeHandler.setSleeper(sleeperOverride);
-      }
-      this.delegateHandler = errorCodeHandler;
+  private static class UnsuccessfulResponseHandler implements HttpUnsuccessfulResponseHandler {
+
+    private static final int HTTP_SC_GONE = 410;
+
+    /** HTTP status code indicating too many requests in a given amount of time. */
+    private static final int HTTP_SC_TOO_MANY_REQUESTS = 429;
+
+    // The set of response codes to log URLs for with a rate limit.
+    private static final ImmutableSet<Integer> RESPONSE_CODES_TO_LOG_WITH_RATE_LIMIT =
+        ImmutableSet.of(HTTP_SC_TOO_MANY_REQUESTS);
+
+    // The set of response codes to log URLs for.
+    private static final ImmutableSet<Integer> RESPONSE_CODES_TO_LOG =
+        ImmutableSet.<Integer>builder()
+            .addAll(RESPONSE_CODES_TO_LOG_WITH_RATE_LIMIT)
+            .add(HTTP_SC_GONE, HttpStatusCodes.STATUS_CODE_SERVICE_UNAVAILABLE)
+            .build();
+
+    // Base implementation of BackOffRequired determining the default set of cases where we'll retry
+    // on unsuccessful HTTP responses; we'll mix in additional retryable response cases on top
+    // of the bases cases defined by this instance.
+    private static final HttpBackOffUnsuccessfulResponseHandler.BackOffRequired BACK_OFF_REQUIRED =
+        response ->
+            response.getStatusCode() == HTTP_SC_TOO_MANY_REQUESTS
+                || HttpBackOffUnsuccessfulResponseHandler.BackOffRequired.ON_SERVER_ERROR
+                    .isRequired(response);
+
+    private final HttpCredentialsAdapter credentials;
+    private final HttpBackOffUnsuccessfulResponseHandler delegate;
+
+    public UnsuccessfulResponseHandler(HttpCredentialsAdapter credentials) {
+      this.credentials = credentials;
+      this.delegate =
+          new HttpBackOffUnsuccessfulResponseHandler(BACKOFF_BUILDER.build())
+              .setBackOffRequired(BACK_OFF_REQUIRED);
     }
 
     @Override
     public boolean handleResponse(HttpRequest request, HttpResponse response, boolean supportsRetry)
         throws IOException {
+      logResponseCode(request, response);
+
       if (credentials != null && credentials.handleResponse(request, response, supportsRetry)) {
         // If credentials decides it can handle it, the return code or message indicated something
         // specific to authentication, and no backoff is desired.
         return true;
       }
 
-      if (delegateHandler.handleResponse(request, response, supportsRetry)) {
+      if (delegate.handleResponse(request, response, supportsRetry)) {
         // Otherwise, we defer to the judgement of our internal backoff handler.
         return true;
       }
```

Comment:
Should we move this check to top of method to catch frequent cases early ?

Reply:
I do not think so, we want to log responses regardless of handling and initialize with credentials if necessary first.